src/charm.py

```diff
@@ -0,0 +1,216 @@
+
+    def __init__(self, *args):
+        """Construct.
+
+        Args:
+            args: Ignore.
+        """
+        super().__init__(*args)
+        self._state = State(self.app, lambda: self.model.get_relation("peer"))
+        self.name = "openldap"
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(
+            self.on.openldap_pebble_ready, self._on_openldap_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.restart_action, self._on_restart)
+        self.framework.observe(
+            self.on.get_admin_password_action, self._on_get_admin_password
+        )
+        self.framework.observe(
+            self.on.load_test_users_action, self._on_load_test_users
+        )
+        self.provider = LDAPProvider(self)
+
+    @log_event_handler(logger)
+    def _on_install(self, event):
+        """Install application.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = MaintenanceStatus("installing OpenLdap")
+
+    @log_event_handler(logger)
+    def _on_openldap_pebble_ready(self, event: ops.PebbleReadyEvent):
+        """Define and start openldap using the Pebble API.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.update(event)
+
+    @log_event_handler(logger)
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Handle configuration changes.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.update(event)
+
+    @log_event_handler(logger)
+    def _on_load_test_users(self, event):
+        """Handle loading of test users and groups.
+
+        Args:
+            event: The `load-test-users` action event.
+
+        Raises:
+            ExecError: In case of error during keytool certificate import
+        """
+        self.unit.status = MaintenanceStatus("Running action.")
+        container = self.unit.get_container(self.name)
+        if not container.can_connect():
+            event.defer()
+            return
+
+        admin_pwd = self._state.bind_password
+        base_dn = self._state.base_dn
+
+        command = [
+            "ldapadd",
+            "-x",
+            "-H",
+            "ldap://localhost:389",
+            "-D",
+            f"cn=admin,{base_dn}",
+            "-w",
+            admin_pwd,
+            "-f",
+            "startup.ldif",
+            "-v",
+        ]
+
+        try:
+            container.exec(
+                command, working_dir="/templates", service_context="openldap"
+            ).wait_output()
+            event.set_results(
+                {"result": "test users and groups successfully added"}
+            )
+            self.unit.status = ActiveStatus()
+        except ExecError as e:
+            logger.error(e.stdout)
```

Comment:
I believe the charm would remain in `MaintenanceStatus` if the action fails here. Should it go back to `ActiveStatus` regardless?

Reply:
Yes it should, thanks.